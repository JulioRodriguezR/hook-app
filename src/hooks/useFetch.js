import { useState, useEffect, useRef } from 'react'

export const useFetch = (url) => {

  const isMounted = useRef(true);
  const [state, setState] = useState({ data: null, loading: true, error: null });

  useEffect(() => {
    return (() => {
      console.log(isMounted);
      isMounted.current = false;
    })
  }, []);

  useEffect(() => {
    setState({ data: null, loading: true, error: null })
    fetch(url)
      .then(resp => resp.json())
      .then(data => {
        if (isMounted.current) {
          console.log('Fetch');
          setState({ loading: false, error: null, data })
        } else {
          console.log('No fetch');
        }
      })
      .catch(() => {
        setState({
          data: null,
          loading: false,
          error: 'Error loading'
        })
      })
  }, [url]);

  return (state);
}
