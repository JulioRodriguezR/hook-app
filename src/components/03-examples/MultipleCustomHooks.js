import React from 'react';

import { useFetch } from '../../hooks/useFetch';
import { useCouter } from '../../hooks/useCouter';

export const MultipleCustomHooks = () => {

  const { counter, increment } = useCouter(1);

  const { data, loading } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
  const { author, quote } = !!data && data[0];

  return (
    <div>
      <h1>Custom Hooks</h1>
      <hr />
      {
        loading
          ?
          (
            <div className="alert alert-info text-center">
              Loading...
            </div>
          )
          :
          (
            <blockquote className="blockquote text-center">
              <div className="border border-2 rounded-0">

                <p className="mb-10">{quote}</p>
                <footer className="blockquote-footer">{author}</footer>

              </div>
              <button className="btn text-center" onClick={increment}>+1</button>
            </blockquote>
          )
      }
    </div>
  )
}
