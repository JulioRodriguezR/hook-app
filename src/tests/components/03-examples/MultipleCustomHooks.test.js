import React from 'react';
import { shallow } from 'enzyme';

import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Test in <MultipleCustomHooks />', () => {

  useCounter.mockReturnValue({ counter: 10, increment: () => {} });

  test('Should show correctly', () => {
    useFetch.mockReturnValue({ data: null, loading: true, error: null });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper).toMatchSnapshot(); 
  });

  test('Should show correct info', () => {
    useFetch.mockReturnValue({
      data: [{ data: null, loading: true, error: null }],
      loading: false,
      error: null
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper.find('alert').exists()).toBe(false); 
    expect(wrapper.find('.mb-10').text().trim()).toBe('Hola Mundo'); 
  });

});