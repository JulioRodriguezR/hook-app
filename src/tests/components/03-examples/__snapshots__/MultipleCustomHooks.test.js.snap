// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <MultipleCustomHooks /> Should show correctly 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
</div>
`;
